module3/homework/advanced_calculator/mathOperation.cpp

```diff
@@ -0,0 +1,158 @@
+#include "mathOperation.hpp"
+#include "errorCodeEnum.hpp"
+#include <cmath>
+#include <cctype>
+#include <iostream>
+#include <vector>
+
+mathOperation::mathOperation(std::string& inputString)
+    : dataString_(inputString), result_(0), firstValue_(0), secondValue_(0),
+      operationCharacter_(""), actualError_(ErrorCode::OK) {
+        removeSpaces();
+        actualError_ =  checkIfBadFormat();
+        actualError_ =  checkIfBadCharacter();
+        actualError_ =  checkIfDivideByZero();
+        actualError_ = checkIfSqrtOfNegativeNumber();
+        actualError_ = checkIfmoduleOfNonIntegerValue();
+        if(actualError_ == ErrorCode::OK) calculateResult();
+
+      };
+
+double mathOperation::getResult() const { return result_; }
+ErrorCode mathOperation::getErrorCode() const { return actualError_; }
+
+
+double mathOperation::add(double &firstValue, double &secondValue) {
+  return firstValue + secondValue;
+}
+double mathOperation::subtract(double &firstValue, double &secondValue) {
+  return firstValue - secondValue;
+}
+double mathOperation::multiplicate(double &firstValue, double &secondValue) {
+  return firstValue * secondValue;
+}
+double mathOperation::divide(double &firstValue, double &secondValue) {
+  return firstValue / secondValue;
+}
+double mathOperation::modulo(double &firstValue, double &secondValue) {
+  return static_cast<double>(static_cast<int>(firstValue)%
+                             static_cast<int>(secondValue));
+}
+double mathOperation::factorial(double &firstValue, double &secondValue) {
+  if(firstValue < 0){
+    return 1;
+  }else return static_cast<double>(std::tgamma(firstValue+1));
+}
+double mathOperation::squareRoot(double &firstValue, double &secondValue) {
+  return pow(firstValue, 1/secondValue);
+}
+double mathOperation::power(double &firstValue, double &secondValue) {
+  return pow(firstValue, secondValue);
+}
+
+void mathOperation::removeSpaces() {
+  dataString_.erase(remove_if(dataString_.begin(), dataString_.end(), isspace),
+                    dataString_.end());
+}
+
+  // #1 group holds complete first value,
+  // #2 group holds what is after dot in first value
+  // #3 group if exists, can hold only '!'
+  // #4 group holds every possible operation sign (with // too)
+  // #5 group is only for one test-case with '//' (Dzieki Mateusz :) :) :) )
+  // #6 group holds complete second value,
+  // #7 group holds what is after dot in second value
+ 
+ErrorCode mathOperation::checkIfBadFormat() {
+
+  const std::regex regexPattern_(
+      "^(-?[0-9]+\\.?([0-9]+)?)(?:(?=!)(!)|(?!!)([\\+\\-\\*\\/\\^\\$%!#@&?;\\[\\]]?(\\\\)?)(-?[0-9]+\\.?([0-9]+)?))$");
```

Comment:
If you split it to 2 regex one for factorial and second for rest.

It will be easier to read.